transports/tls/src/upgrade.rs

```diff
@@ -118,6 +123,63 @@ where
     }
 }
 
+impl<C> InboundSecurityUpgrade<C> for Config
+where
+    C: AsyncRead + AsyncWrite + Send + Unpin + 'static,
+{
+    type Output = TlsStream<C>;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_inbound(self, socket: C, _: Self::Info, _: Option<PeerId>) -> Self::Future {
+        async move {
+            let stream = futures_rustls::TlsAcceptor::from(Arc::new(self.server))
+                .accept(socket)
+                .await
+                .map_err(UpgradeError::ServerUpgrade)?;
+
+            let expected = extract_single_certificate(stream.get_ref().1)?.peer_id();
+
+            Ok((expected, stream.into()))
+        }
+        .boxed()
+    }
+}
+
+impl<C> OutboundSecurityUpgrade<C> for Config
+where
+    C: AsyncRead + AsyncWrite + Send + Unpin + 'static,
+{
+    type Output = TlsStream<C>;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_outbound(self, socket: C, _: Self::Info, peer_id: Option<PeerId>) -> Self::Future {
+        async move {
+            // Spec: In order to keep this flexibility for future versions, clients that only support
+            // the version of the handshake defined in this document MUST NOT send any value in the
+            // Server Name Indication.
+            // Setting `ServerName` to unspecified will disable the use of the SNI extension.
+            let name = ServerName::IpAddress(IpAddr::V4(Ipv4Addr::UNSPECIFIED));
+
+            let stream = futures_rustls::TlsConnector::from(Arc::new(self.client))
+                .connect(name, socket)
+                .await
+                .map_err(UpgradeError::ClientUpgrade)?;
+
+            let expected = extract_single_certificate(stream.get_ref().1)?.peer_id();
+
+            match peer_id {
+                Some(found) if found != expected => {
+                    Err(UpgradeError::PeerIdMismatch { expected, found })
+                }
+                _ => Ok((expected, stream.into())),
+            }
```

Comment:
We don't need to do this. Instead, we can make the client and server config ad-hoc in here and pass the expected `PeerId` into `crate::make_client_config`.

Reply:
Thanks, I'll get into this tomorrow.